### ISSUE TITLE
Propagator missing edge leads to Phi incorrectly constant folded

### ISSUE THREAD

**Issue by c0d1f1ed** [KhronosGroup/SPIRV-Tools#4462]
SwiftShader [fails to pass dEQP-VK.graphicsfuzz.stable-rects-vec4-clamp-conditional-min-mix](https://issuetracker.google.com/195422130) due to a bug in spirv-opt.

The minimal repro is this GLSL code:
```
float x = clamp(0.5, 0.5, (gl_FragCoord.x < 0.0) ? -1.0 : (gl_FragCoord.x >= 0.0 ? (mix(1.0, 0.5, (gl_FragCoord.x >= 0.0))) : 1.0, 0.5));
outColor = vec4(x, 0.5, 0.0, 1.0);
```
It results in x = -1.0, instead of the expected 0.5.

Before the CPP pass, the relevant SPIR-V looks like this:
```
%9 = OpConstant %6 0.5
%18 = OpConstant %6 0
%24 = OpConstant %6 -1
%32 = OpConstant %6 1
...
%17 = OpLoad %6 %16
%20 = OpFOrdLessThan %19 %17 %18
OpSelectionMerge %23 None
OpBranchConditional %20 %22 %25
%22 = OpLabel
OpBranch %23
%25 = OpLabel
OpBranch %31
%31 = OpLabel
OpBranch %23
%23 = OpLabel
%44 = OpPhi %6 %24 %22 %9 %31
%39 = OpExtInst %6 %1 FClamp %9 %9 %44
%43 = OpCompositeConstruct %10 %39 %9 %18 %32
```
After CPP, the only change is:
```
%43 = OpCompositeConstruct %10 %24 %9 %18 %32
```
Explaining why this happens is best done backwards. `FoldClamp1()` receives { 0.5, 0.5, -1.0 }, which makes it return -1.0. The spec states that "The resulting value is undefined if minVal > maxVal." so this behavior is not wrong.

The reason the third operand for the Clamp is -1.0, is because `SSAPropagator::IsPhiArgExecutable()` returns false for the %31 -> %23 edge and this makes `CCPPass::VisitPhi()` only look at the first var, and thus it assumes it can only ever produce -1.0.

Further analysis is needed to root cause why this edge is not considered executable, but it appears that block %31 only being indirectly reachable through %25 has something to do with it. Running `BlockMergePass` before `CCPPass` works around the issue, but it's not clear that it would address every other edge case.

**Comment by dnovillo** [comment]
@c0d1f1ed with current top-of-main, I'm getting the final `OpStore` to `%x` folded into `0.5f`.  Isn't that the expected value?

```
 71          %41 = OpExtInst %float %1 FClamp %float_0_5 %float_0_5 %40
 72                OpStore %x %41
 73          %42 = OpLoad %float %x
```

becomes

```
 71          %41 = OpExtInst %float %1 FClamp %float_0_5 %float_0_5 %40
 72                OpStore %x %float_0_5
 73          %42 = OpLoad %float %x
```

precisely because, as you point out, the value of `%40` doesn't really affect the result from `FClamp`.

**Comment by c0d1f1ed** [comment]
I updated SwiftShader's subtree of SPIRV-Tools to the latest revision today, but I'm still seeing the original issue. I hacked up our `VulkanBenchmarks` to reproduce it easily: https://swiftshader-review.googlesource.com/c/SwiftShader/+/60288. It renders a green triangle, instead of the expected yellow. I've attached the pre- and post-optimization SPIR-V. The latter puts -1 in the `x` component. Applying the [workaround](https://swiftshader-review.googlesource.com/c/SwiftShader/+/56208) still fixes it.

[post-opt.txt](https://github.com/KhronosGroup/SPIRV-Tools/files/7627708/post-opt.txt)
[pre-opt.txt](https://github.com/KhronosGroup/SPIRV-Tools/files/7627709/pre-opt.txt)

**Comment by s-perron** [comment]
To reproduce what @c0d1f1ed sees, you need to run all optimizations (`-O`) on the spir-v provided.  If you just run `ccp`, then you get what @dnovillo mentioned.

**Comment by s-perron** [comment]
CCP is doing something wrong

Before CCP:

```
%44 = OpPhi %8 %18 %28 %10 %33
%41 = OpExtInst %8 %1 FClamp %10 %10 %44
%43 = OpCompositeConstruct %11 %41 %10 %16 %19
OpStore %4 %43
```

After CCP

```
%45 = OpConstantComposite %11 %18 %10 %16 %19
...
%44 = OpPhi %8 %18 %28 %10 %33
%41 = OpExtInst %8 %1 FClamp %10 %10 %44
%43 = OpCompositeConstruct %11 %10 %10 %16 %19
OpStore %4 %45
```

Notice that for the def of `%43` is propagated `%10` (0.5), but for the store is has `%18` (-1) for the first element.

**Comment by dnovillo** [comment]
Thanks.  I've found the root cause and fixed it in #4651.

The problem is with `FClamp` being folded to different constant values in two separate evaluations.  Both fold results are valid, but CCP was not correctly detecting that after the second fold operation returned a different value, it should've considered the `FClamp` instruction to be varying.  So, subsequent uses of its result were left using the original folded value.

In CCP, the constant lattice can only be traversed top down.  An instruction that folds to two different constant values should be considered varying (otherwise, the algorithm could enter in a non-convergence state).
